multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/ReplayActionTest.java

```diff
@@ -0,0 +1,161 @@
+        WorkflowRun b2 = (WorkflowRun) b1.getAction(ReplayAction.class).run("node {checkout scm; echo readFile('file').toUpperCase()}").get();
+        assertEquals(2, b2.number);
+        r.assertLogContains("INITIAL CONTENT", b2);
+    }
+
+    @Test public void permissions() throws Exception {
+        File clones = tmp.newFolder();
+        sampleRepo.init();
+        sampleRepo.write("Jenkinsfile", "");
+        sampleRepo.git("add", "Jenkinsfile");
+        sampleRepo.git("commit", "--all", "--message=init");
+        sampleRepo.git("clone", ".", new File(clones, "one").getAbsolutePath());
+        r.jenkins.setSecurityRealm(r.createDummySecurityRealm());
+        ProjectMatrixAuthorizationStrategy pmas = new ProjectMatrixAuthorizationStrategy();
+        pmas.add(Jenkins.ADMINISTER, "admin");
+        pmas.add(Jenkins.READ, "dev1");
+        pmas.add(Jenkins.READ, "dev2");
+        pmas.add(Jenkins.READ, "dev3");
+        r.jenkins.setAuthorizationStrategy(pmas);
+        OrganizationFolder top = r.jenkins.createProject(OrganizationFolder.class, "top");
+        Map<Permission, Set<String>> perms = new HashMap<Permission, Set<String>>();
+        perms.put(Item.CONFIGURE, Collections.singleton("dev1")); // implies REPLAY
+        perms.put(ReplayAction.REPLAY, Collections.singleton("dev2"));
+        perms.put(Item.BUILD, Collections.singleton("dev3")); // does not imply REPLAY
+        top.addProperty(new AuthorizationMatrixProperty(perms));
+        top.getNavigators().add(new GitDirectorySCMNavigator(clones.getAbsolutePath()));
+        top.scheduleBuild2(0).getFuture().get();
+        top.getComputation().writeWholeLogTo(System.out);
+        assertEquals(1, top.getItems().size());
+        MultiBranchProject<?,?> one = top.getItem("one");
+        r.waitUntilNoActivity();
+        WorkflowJob p = WorkflowMultiBranchProjectTest.findBranchProject((WorkflowMultiBranchProject) one, "master");
+        WorkflowRun b1 = p.getLastBuild();
+        assertEquals(1, b1.getNumber());
+        assertTrue(canReplay(b1, "admin"));
+        assertTrue(canReplay(b1, "dev1"));
+        assertTrue(canReplay(b1, "dev2"));
+        assertFalse(canReplay(b1, "dev3"));
+        p = r.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("", false));
+        b1 = p.scheduleBuild2(0).get();
+        assertTrue(canReplay(b1, "admin"));
+        assertFalse("not sandboxed, so only safe for admins", canReplay(b1, "dev1"));
+        assertFalse(canReplay(b1, "dev2"));
+        assertFalse(canReplay(b1, "dev3"));
+    }
+    private static boolean canReplay(WorkflowRun b, String user) {
+        final ReplayAction a = b.getAction(ReplayAction.class);
+        return ACL.impersonate(User.get(user).impersonate(), new NotReallyRoleSensitiveCallable<Boolean,RuntimeException>() {
+            @Override public Boolean call() throws RuntimeException {
+                return a.isEnabled();
+            }
+        });
+    }
+
+}
```

Comment:
Again... awesome tests ... great detail, coverage etc, but reading it is like wading through sweet treacle :smile: